backstory/src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java

```diff
@@ -0,0 +1,229 @@
+    // form in the HTML will connect to the Blobstore URL, which encodes the images and then
+    // redirects the request to this Url. In this doPost, the images are analyzed with Vision API,
+    // and the analysis along with the image URL will be sent to datastore for permanent storage.
+
+    // Get the input from the form.
+    // Get the URL of the image that the user uploaded to Blobstore.
+    final String imageUrl = getUploadedFileUrl(request, "image-upload");
+
+    // Get the raw byte array representing the image from Blobstore
+    final byte[] bytes = getBlobBytes(request, "image-upload");
+
+    // Gets the full label information from the image byte array by calling Vision API.
+    List<EntityAnnotation> labels = detectLabelsBytes(bytes);
+    Gson gson = new Gson();
+    Text labelsJsonArray = new Text(gson.toJson(labels));
+
+    // Add the input to datastore
+    Entity analyzedImageEntity = new Entity("analyzed-image");
+    analyzedImageEntity.setProperty("imageUrl", imageUrl);
+    analyzedImageEntity.setProperty("labelsJsonArray", labelsJsonArray);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(analyzedImageEntity);
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/visionUploadPrototype/visionDemo.html#image-upload");
+  }
+
+  // Detects labels in the image specified by the image byte data by calling the Vision API.
+  private List<EntityAnnotation> detectLabelsBytes(byte[] bytes) throws IOException {
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    List<EntityAnnotation> labels;
+
+    Image img = Image.newBuilder().setContent(ByteString.copyFrom(bytes)).build();
+    Feature feat = Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder().addFeatures(feat).setImage(img).build();
+    requests.add(request);
+
+    // Initialize client that will be used to send requests. This client only needs to be created
+    // once, and can be reused for multiple requests. After completing all of the requests the
+    // client will be automatically closed, as it is called within the try.
+    try (ImageAnnotatorClient client = ImageAnnotatorClient.create()) {
+      BatchAnnotateImagesResponse response = client.batchAnnotateImages(requests);
+      List<AnnotateImageResponse> responses = response.getResponsesList();
+
+      // There is only one image in the batch response (only supports uploading one image at a time
+      // right now)
+      AnnotateImageResponse res = responses.get(0);
+
+      if (res.hasError()) {
+        System.out.format("Error: %s%n", res.getError().getMessage());
+        return new ArrayList<EntityAnnotation>();
+      }
+
+      // For full list of available annotations, see http://g.co/cloud/vision/docs
+      labels = res.getLabelAnnotationsList();
+    }
+
+    return labels;
+  }
```

Comment:
Tests?

Reply:
There will be tests for the final VisionManager, but remember that we agreed the prototypes do not need tests?

Reply:
Will some of this demo functionality be used in the final classes? If so, I think it would be valuable to extract that logic into those final classes, to needing to duplicate that code when implementing the final classes.